pkg/policy/verifier.go

```diff
@@ -0,0 +1,210 @@
+// is customizable by callers that's suitable for their execution
+// environment.
+type WarningWriter func(string, ...interface{})
+
+// Compile turns a Verification into an executable Verifier.
+// Any compilation errors are returned here.
+func Compile(ctx context.Context, v Verification, ww WarningWriter) (Verifier, error) {
+	if err := v.Validate(ctx); err != nil {
+		return nil, err
+	}
+
+	ipc, err := gather(ctx, v, ww)
+	if err != nil {
+		// This should never hit for validated policies.
+		return nil, err
+	}
+
+	return &impl{
+		verification: v,
+		ipc:          ipc,
+		ww:           ww,
+	}, nil
+}
+
+func gather(ctx context.Context, v Verification, ww WarningWriter) (*config.ImagePolicyConfig, error) {
+	pol := *v.Policies
+	ipc := &config.ImagePolicyConfig{
+		Policies: make(map[string]webhookcip.ClusterImagePolicy, len(pol)),
+	}
+
+	for i, p := range pol {
+		var content string
+		switch {
+		case p.Data != "":
+			content = p.Data
+
+		case p.Path != "":
+			raw, err := os.ReadFile(p.Path)
+			if err != nil {
+				return nil, err
+			}
+			content = string(raw)
+
+		case p.URL != "":
+			resp, err := http.Get(p.URL)
+			if err != nil {
+				return nil, err
+			}
+			defer resp.Body.Close()
+			raw, err := io.ReadAll(resp.Body)
+			if err != nil {
+				return nil, err
+			}
+			content = string(raw)
+
+		default:
+			// This should never happen for a validated policy.
+			return nil, fmt.Errorf("unsupported policy shape: %v", p)
+		}
+
+		l, warns, err := ParseClusterImagePolicies(ctx, content)
+		if err != nil {
+			// This path should be unreachable, since we already parse
+			// things during compilation.
+			return nil, fmt.Errorf("parsing policies: %w", err)
+		}
+		if warns != nil {
+			ww("policy %d: %v", i, warns)
+		}
+
+		// TODO(mattmoor): Add additional checks for unsupported things,
+		// like Match, IncludeSpec, etc.
+
+		for _, cip := range l {
+			cip.SetDefaults(ctx)
+			if _, ok := ipc.Policies[cip.Name]; ok {
+				ww("duplicate policy named %q, skipping", cip.Name)
+				continue
+			}
+			// We need to roundtrip the policy through JSON here because
+			// the compiled policy expects to be decoded from JSON and only
+			// sets up certain fields when being unmarshalled from JSON, so
+			// things like keyful verification only work when we roundtrip
+			// through JSON.
+			var compiled webhookcip.ClusterImagePolicy
+			if err := convert(webhookcip.ConvertClusterImagePolicyV1alpha1ToWebhook(cip), &compiled); err != nil {
+				ww("roundtripping policy %v", err)
+				continue
+			}
+			ipc.Policies[cip.Name] = compiled
+		}
+	}
+
+	return ipc, nil
+}
+
+type impl struct {
+	verification Verification
+
+	ipc *config.ImagePolicyConfig
+	ww  WarningWriter
+}
+
+// Check that impl implements Verifier
+var _ Verifier = (*impl)(nil)
+
+// Verify implements Verifier
+func (i *impl) Verify(ctx context.Context, ref name.Reference, kc authn.Keychain) error {
+	tm := getTypeMeta(ctx)
```

Comment:
We should document that some things are controlled by context fields that you can set with IncludeMeta, IncludeType because they are also part of the interface though indirectly as implemented.

Reply:
Yeah, good point.  I added some of this after documenting most of this stuff to support `policy-tester` with policies providing `match:`.  I'll add this to the interface in the next push.